### PR TITLE
loop orders only once on to_json

### PR DESCRIPTION
**Pull request by stash86**
this might not matter to all users, but as someone who fetches trades' data regularly over API for my own dashboard, having each trades looping over orders up to 9 times isn't ideal.

AI was used to found the issue, edits were done manually

This is a summary from AI
```
Current passes:
- select_filled_or_open_orders() — 1 pass
- date_entry_fill_utc property → select_filled_orders(entry_side) — 1 pass
- stoploss_last_update_utc property → has_open_sl_orders check — 1 pass
- stoploss_last_update_utc property → open_sl_orders property — 1 pass
- stoploss_last_update_utc accessed again for _timestamp key — 1 pass
- nr_of_successful_entries property → _count_filled_orders(entry_side) — 1 pass
- nr_of_successful_exits property → _count_filled_orders(exit_side) — 1 pass
- has_open_orders property — 1 short-circuit pass

Total: 7-9 passes per to_json() call
```

### CODE REVIEW

File: freqtrade/persistence/trade_model.py [hunk 1]
```diff
@@ -712,16 +743,8 @@ def to_json(self, minified: bool = False) -> dict[str, Any]:
...
-                if self.close_date
-                else None
-            ),
-            "trade_duration": (
-                int((self.close_date_utc - self.open_date_utc).total_seconds() // 60)
-                if self.close_date
-                else None
-            ),
+            "trade_duration_s": trade_duration_s,
+            "trade_duration": int(trade_duration_s // 60) if trade_duration_s else None,
```
Comment (xmatthias) [inline freqtrade/persistence/trade_model.py]: You'll have to compare to `is not None` here - otherwise same second trades will yield None instead of 0 (as 0 is "falsy").

It's also inconsistent with trade_duration_s - as that'll be passed through as is.

File: freqtrade/persistence/trade_model.py [hunk 2]
```diff
@@ -666,8 +666,41 @@ def to_json(self, minified: bool = False) -> dict[str, Any]:
...
+        open_sl_orders_ts = []
+        for order in filled_or_open_orders:
+            if order.ft_is_open:
+                if order.ft_order_side not in ["stoploss"]:
+                    open_orders_wo_sl.append(order)
+                else:
+                    open_sl_orders.append(order)
+                    open_sl_orders_ts.append(order.order_date_utc)
+            else:
+                if order.filled and order.status in NON_OPEN_EXCHANGE_STATES:
```
Comment (xmatthias) [inline freqtrade/persistence/trade_model.py]: isn't this condition already guaranteed by `select_filled_or_open_orders()` ? 

so we might be able to reduce the if nesting by a level here.

**Review by xmatthias — CHANGES_REQUESTED** [review]
I'm ... torn here. 

It's a marginal performance improvement (as in other PR's, not sure it's actually measurable in a real instance) - at the risk of "implementation deviation".

if we change the logic within one of the existing properties for whatever reason in the future, `to_json` will silently deviate / drift - as the logic is re-implemented here (Without a test ensuring they don't deviate).
Now i'm not sure on the value of such a test, either (so don't treat this as "we need a test"), as it'll depend on the change (is it a regular change, or covering some odd edge-case?) - so even an explicit test might not be capturing the deviation here.


---

i fear that the CI failure is a valid regression - as it the test case now seems to have either inf or nan as part of it's output - which didn't happen with previous logic.

While it's not part of a regular pytest run - it's still a test - though a real, valid "run a valid backtest" one ...

**Comment by stash86** [comment]
Your future missed behavior change is valid, so it's up to you if you think it's worth being merged.

I asked Codex to create a script to test the old vs new implementation

<details><summary>Details</summary>
<p>



```py
#!/usr/bin/env python3
"""Benchmark the old and current LocalTrade.to_json implementations.

Run from the repository root, for example:

    python scripts/benchmark_trade_to_json.py
    python scripts/benchmark_trade_to_json.py --trades 1 10 100 --orders 10
    python scripts/benchmark_trade_to_json.py --orders 1 10 100 1000 --repeat 9
    python scripts/benchmark_trade_to_json.py --minified --iterations 10000

The benchmark uses in-memory LocalTrade/Order objects. It does not need a database,
pytest, exchange credentials, or network access.
"""

from __future__ import annotations

import argparse
import gc
import statistics
import timeit
from collections.abc import Callable
from datetime import datetime, timedelta
from functools import partial
from typing import Any

from freqtrade.constants import DATETIME_PRINT_FORMAT
from freqtrade.persistence.trade_model import LocalTrade, Order
from freqtrade.util import dt_ts_none


def old_to_json(self: LocalTrade, minified: bool = False) -> dict[str, Any]:
    """LocalTrade.to_json before the order-loop optimization."""
    filled_or_open_orders = self.select_filled_or_open_orders()
    orders_json = [order.to_json(self.entry_side, minified) for order in filled_or_open_orders]

    return {
        "trade_id": self.id,
        "pair": self.pair,
        "base_currency": self.safe_base_currency,
        "quote_currency": self.safe_quote_currency,
        "is_open": self.is_open,
        "exchange": self.exchange,
        "amount": round(self.amount, 8),
        "amount_requested": round(self.amount_requested, 8) if self.amount_requested else None,
        "stake_amount": round(self.stake_amount, 8),
        "max_stake_amount": round(self.max_stake_amount, 8) if self.max_stake_amount else None,
        "strategy": self.strategy,
        "enter_tag": self.enter_tag,
        "timeframe": self.timeframe,
        "fee_open": self.fee_open,
        "fee_open_cost": self.fee_open_cost,
        "fee_open_currency": self.fee_open_currency,
        "fee_close": self.fee_close,
        "fee_close_cost": self.fee_close_cost,
        "fee_close_currency": self.fee_close_currency,
        "open_date": self.open_date.strftime(DATETIME_PRINT_FORMAT),
        "open_timestamp": dt_ts_none(self.open_date_utc),
        "open_fill_date": (
            self.date_entry_fill_utc.strftime(DATETIME_PRINT_FORMAT)
            if self.date_entry_fill_utc
            else None
        ),
        "open_fill_timestamp": dt_ts_none(self.date_entry_fill_utc),
        "open_rate": self.open_rate,
        "open_rate_requested": self.open_rate_requested,
        "open_trade_value": round(self.open_trade_value, 8),
        "close_date": self.close_date.strftime(DATETIME_PRINT_FORMAT) if self.close_date else None,
        "close_timestamp": dt_ts_none(self.close_date_utc),
        "realized_profit": self.realized_profit or 0.0,
        "realized_profit_ratio": self.close_profit or None,
        "close_rate": self.close_rate,
        "close_rate_requested": self.close_rate_requested,
        "close_profit": self.close_profit,
        "close_profit_pct": round(self.close_profit * 100, 2) if self.close_profit else None,
        "close_profit_abs": self.close_profit_abs,
        "trade_duration_s": (
            int((self.close_date_utc - self.open_date_utc).total_seconds())
            if self.close_date
            else None
        ),
        "trade_duration": (
            int((self.close_date_utc - self.open_date_utc).total_seconds() // 60)
            if self.close_date
            else None
        ),
        "profit_ratio": self.close_profit,
        "profit_pct": round(self.close_profit * 100, 2) if self.close_profit else None,
        "profit_abs": self.close_profit_abs,
        "exit_reason": self.exit_reason,
        "exit_order_status": self.exit_order_status,
        "stop_loss_abs": self.stop_loss,
        "stop_loss_ratio": self.stop_loss_pct if self.stop_loss_pct else None,
        "stop_loss_pct": (self.stop_loss_pct * 100) if self.stop_loss_pct else None,
        "stoploss_last_update": (
            self.stoploss_last_update_utc.strftime(DATETIME_PRINT_FORMAT)
            if self.stoploss_last_update_utc
            else None
        ),
        "stoploss_last_update_timestamp": dt_ts_none(self.stoploss_last_update_utc),
        "initial_stop_loss_abs": self.initial_stop_loss,
        "initial_stop_loss_ratio": (
            self.initial_stop_loss_pct if self.initial_stop_loss_pct else None
        ),
        "initial_stop_loss_pct": (
            self.initial_stop_loss_pct * 100 if self.initial_stop_loss_pct else None
        ),
        "min_rate": self.min_rate,
        "max_rate": self.max_rate,
        "leverage": self.leverage,
        "interest_rate": self.interest_rate,
        "liquidation_price": self.liquidation_price,
        "is_short": self.is_short,
        "trading_mode": self.trading_mode,
        "funding_fees": self.funding_fees,
        "amount_precision": self.amount_precision,
        "price_precision": self.price_precision,
        "precision_mode": self.precision_mode,
        "precision_mode_price": self.precision_mode_price,
        "contract_size": self.contract_size,
        "nr_of_successful_entries": self.nr_of_successful_entries,
        "nr_of_successful_exits": self.nr_of_successful_exits,
        "has_open_orders": self.has_open_orders,
        "orders": orders_json,
    }


def make_trade(order_count: int) -> LocalTrade:
    """Create a deterministic trade containing a representative order mix."""
    base_date = datetime(2026, 1, 1, 12, 0)
    trade = LocalTrade(
        pair="ETH/USDT",
        stake_amount=100.0,
        max_stake_amount=100.0,
        amount=0.05,
        amount_requested=0.05,
        fee_open=0.001,
        fee_close=0.001,
        is_open=True,
        open_date=base_date,
        open_rate=2000.0,
        exchange="benchmark",
        strategy="BenchmarkStrategy",
        timeframe=5,
    )

    # Repeating four-state mix: filled entry, filled exit, open entry, open stop-loss.
    for index in range(order_count):
        state = index % 4
        is_open = state >= 2
        side = ("buy", "sell", "buy", "stoploss")[state]
        order_date = base_date + timedelta(seconds=index)
        filled = 0.05 if not is_open else 0.0
        trade.orders.append(
            Order(
                ft_order_side=side,
                ft_pair=trade.pair,
                ft_is_open=is_open,
                ft_amount=0.05,
                ft_price=2000.0,
                order_id=f"benchmark-{index}",
                status="open" if is_open else "closed",
                side=side,
                order_type="stoploss" if side == "stoploss" else "limit",
                price=2000.0,
                average=2000.0 if filled else None,
                amount=0.05,
                filled=filled,
                remaining=0.05 - filled,
                cost=filled * 2000.0,
                order_date=order_date,
                order_filled_date=order_date if filled else None,
            )
        )

    return trade


def old_batch_to_json(trades: list[LocalTrade], minified: bool) -> list[dict[str, Any]]:
    return [old_to_json(trade, minified) for trade in trades]


def new_batch_to_json(trades: list[LocalTrade], minified: bool) -> list[dict[str, Any]]:
    return [trade.to_json(minified) for trade in trades]


def measure(function: Callable[[], object], iterations: int) -> float:
    """Return total seconds for the requested number of calls."""
    return timeit.Timer(function).timeit(number=iterations)


def calibrated_iterations(function: Callable[[], object], target_seconds: float) -> int:
    """Choose enough iterations to make one sample approximately target_seconds long."""
    iterations = 1
    elapsed = 0.0
    while elapsed < 0.05 and iterations < 1_000_000:
        elapsed = measure(function, iterations)
        if elapsed < 0.05:
            iterations *= 10
    return max(1, min(1_000_000, round(iterations * target_seconds / elapsed)))


def format_duration(seconds: float) -> str:
    magnitude = abs(seconds)
    if magnitude < 1e-6:
        return f"{seconds * 1e9:.1f} ns"
    if magnitude < 1e-3:
        return f"{seconds * 1e6:.2f} us"
    if magnitude < 1:
        return f"{seconds * 1e3:.2f} ms"
    return f"{seconds:.2f} s"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--orders",
        type=int,
        nargs="+",
        default=[1, 10, 100],
        help="number of orders in each trade (default: 1 10 100)",
    )
    parser.add_argument(
        "--trades",
        type=int,
        nargs="+",
        default=[1],
        help="number of trades serialized in each batch (default: 1)",
    )
    parser.add_argument(
        "--repeat",
        type=int,
        default=7,
        help="timing samples per implementation (default: 7)",
    )
    parser.add_argument(
        "--iterations",
        type=int,
        help="batches per sample; by default this is calibrated for each case",
    )
    parser.add_argument(
        "--target-time",
        type=float,
        default=0.25,
        help="target seconds per auto-calibrated sample (default: 0.25)",
    )
    parser.add_argument(
        "--minified",
        action="store_true",
        help="pass minified=True to both implementations",
    )
    args = parser.parse_args()
    if any(count < 0 for count in args.orders):
        parser.error("--orders values must be non-negative")
    if any(count < 1 for count in args.trades):
        parser.error("--trades values must be at least 1")
    if args.repeat < 1:
        parser.error("--repeat must be at least 1")
    if args.iterations is not None and args.iterations < 1:
        parser.error("--iterations must be at least 1")
    if args.target_time <= 0:
        parser.error("--target-time must be positive")
    return args


def main() -> None:
    args = parse_args()
    print(f"Mode: {'minified' if args.minified else 'full'} JSON, median of {args.repeat} samples")
    print(
        f"{'orders/trade':>12}  {'trades':>8}  {'batches/sample':>14}  "
        f"{'old total':>12}  {'new total':>12}  {'time saved':>12}  "
        f"{'saved':>8}  {'speedup':>8}"
    )

    gc_was_enabled = gc.isenabled()
    gc.disable()
    try:
        for order_count in args.orders:
            for trade_count in args.trades:
                trades = [make_trade(order_count) for _ in range(trade_count)]
                old_call = partial(old_batch_to_json, trades, args.minified)
                new_call = partial(new_batch_to_json, trades, args.minified)

                old_result = old_call()
                new_result = new_call()
                if old_result != new_result:
                    raise RuntimeError(
                        f"old/new output differs for {trade_count} trades with "
                        f"{order_count} orders each"
                    )

                work_per_batch = trade_count * max(1, order_count)
                warmup_runs = max(1, min(100, 10_000 // work_per_batch))
                for _ in range(warmup_runs):
                    old_call()
                    new_call()

                iterations = args.iterations or calibrated_iterations(new_call, args.target_time)
                old_samples: list[float] = []
                new_samples: list[float] = []
                for sample in range(args.repeat):
                    # Alternate measurement order to reduce thermal/frequency bias.
                    if sample % 2:
                        new_samples.append(measure(new_call, iterations))
                        old_samples.append(measure(old_call, iterations))
                    else:
                        old_samples.append(measure(old_call, iterations))
                        new_samples.append(measure(new_call, iterations))

                old_batch_time = statistics.median(old_samples) / iterations
                new_batch_time = statistics.median(new_samples) / iterations
                saved_batch_time = old_batch_time - new_batch_time
                saved_percent = saved_batch_time / old_batch_time * 100
                speedup = old_batch_time / new_batch_time
                print(
                    f"{order_count:12,d}  {trade_count:8,d}  {iterations:14,d}  "
                    f"{format_duration(old_batch_time):>12}  "
                    f"{format_duration(new_batch_time):>12}  "
                    f"{format_duration(saved_batch_time):>12}  "
                    f"{saved_percent:7.2f}%  {speedup:7.2f}x"
                )
    finally:
        if gc_was_enabled:
            gc.enable()


if __name__ == "__main__":
    main()
```

</p>
</details> 

I ran this command `python3 benchmark_trade_to_json.py --trades 1 10 100 --orders 1 10 50`
Got this result
```
Mode: full JSON, median of 7 samples
orders/trade    trades  batches/sample     old total     new total    time saved     saved   speedup
           1         1          16,606      19.87 us      15.03 us       4.84 us    24.37%     1.32x
           1        10           1,638     200.44 us     150.96 us      49.48 us    24.69%     1.33x
           1       100             168       1.99 ms       1.50 ms     489.18 us    24.60%     1.33x
          10         1           3,014     106.94 us      81.64 us      25.30 us    23.65%     1.31x
          10        10             309       1.07 ms     815.28 us     250.35 us    23.49%     1.31x
          10       100              30      10.80 ms       8.23 ms       2.57 ms    23.83%     1.31x
          50         1             678     480.28 us     368.54 us     111.74 us    23.26%     1.30x
          50        10              67       4.81 ms       3.70 ms       1.12 ms    23.16%     1.30x
          50       100               6      51.49 ms      38.31 ms      13.18 ms    25.59%     1.34x
```

**Comment by xmatthias** [comment]
I'm not arguing the change per se - just ... highlighting it's potential.

As i said - i also don't think explicit test coverage of that makes much sense. On the other hand, i also don't expect these to change necessarily.

**Review by xmatthias — APPROVED** [review]
LGTM 👍
